### PR TITLE
Add `unit` property and `as_unit` method to `Timestamp`, `Timedelta` and `NaTType`

### LINKED ISSUE #797: "Timestamp" has no attribute "as_unit" [attr-defined]
I think we are missing the `.as_unit` method added to pd.Timestamp in 2.0

### LINKED ISSUE #859: Timestamp.as_unit
**Describe the bug**
Method `as_unit` not defined on `Timestamp`

**To Reproduce**
1. Provide a minimal runnable `pandas` example that is not properly checked by the stubs.
`pd.Timestamp(2023, 1, 1).as_unit("ns")`
2. Indicate which type checker you are using (`mypy` or  `pyright`).
mypy
3. Show the error message received from that type checker while checking your example.
error: "Timestamp" has no attribute "as_unit"  [attr-defined]

**Please complete the following information:**
 - OS: Linux
 - OS Fedora 7
 - python version: 3.11
 - version of type checker: 1.8
 - version of installed `pandas-stubs`=2.1.4

### PR DESCRIPTION
**Pull request by skatsuta**
- [x] Closes #797
- [x] Closes #859 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

Add `unit` property and `as_unit` method to `Timestamp`, `Timedelta` and `NaTType`.

Adding them to `DatetimeIndex`, `TimedeltaIndex` and `Series.dt` will be done in #865.

### CODE REVIEW

File: pandas-stubs/_libs/tslibs/nattype.pyi [hunk 1]
```diff
@@ -121,3 +125,8 @@ class NaTType:
     __le__: _NatComparison
     __gt__: _NatComparison
     __ge__: _NatComparison
+    @property
+    def unit(self) -> str: ...
+    def as_unit(
+        self, unit: Literal["s", "ms", "us", "ns"], round_ok: bool = ...
+    ) -> Self: ...
```
Comment (Dr-Irv) [inline pandas-stubs/_libs/tslibs/nattype.pyi:132]: Can you create a new type in `_typing.pyi` called `TimeUnit` that is `Literal["s", "ms", "us", "ns"]` and use that as the return type of `unit` and as the argument to `as_unit()` ?
Also replace wherever we have in the stubs that `Literal` and use `TimeUnit` instead.

Note that `TimeUnit` is now in `pandas/_typing.py` .

File: tests/test_scalars.py [hunk 2]
```diff
@@ -517,6 +517,7 @@ def test_timedelta_properties_methods() -> None:
     check(assert_type(td.value, int), int)
     check(assert_type(td.resolution_string, str), str)
     check(assert_type(td.components, Components), Components)
+    check(assert_type(td.unit, str), str)
```
Comment (Dr-Irv) [inline tests/test_scalars.py:520]: Since `TimeUnit` is in `pandas`, you could import that and use it here in the `assert_type` test.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Suggest using `TimeUnit` that is now in pandas.

**Comment by Dr-Irv** [comment]
Closing in favor of #864

**Review by skatsuta — COMMENTED** [review]
@Dr-Irv Thanks for your suggestion.

Defined the `TimeUnit` type alias and used it throughout the stubs.

Also rebased to resolve conflicts with the main branch, so apologize if additional changes are not clear.

Will update #865 to use `TimeUnit` as well once this PR gets merged.

**Review by Dr-Irv — APPROVED** [review]
thanks @skatsuta

**Pull request merged**